libmamba/src/core/history.cpp

```diff
@@ -325,4 +325,183 @@ namespace mamba
+            std::string channel;
+            std::size_t pos_0 = s.rfind("/");
+            if (pos_0 != std::string::npos)
+            {
+                std::string s_begin = s.substr(0, pos_0);
+                std::string s_end = s.substr(pos_0 + 1, s.size());
+                std::size_t pos = s_begin.rfind("/");
+                if (pos != std::string::npos)
+                {
+                    channel = s_begin.substr(pos + 1, s_begin.size());
+                }
+                else
+                {
+                    channel = s_begin;
+                }
+                s_pkg = std::get<1>(util::rsplit_once(s_end, "::"));
+            }
+            else
+            {
+                channel = std::get<0>(util::split_once(s, "::"));
+                s_pkg = std::get<1>(util::rsplit_once(s, "::"));
+            }
+
+            std::size_t pos_1 = s_pkg.rfind("-");
+            std::string s_pkg_ = s_pkg.substr(0, pos_1);
+            std::string build_string = s_pkg.substr(pos_1 + 1, s_pkg.size());
+
+            std::size_t pos_2 = s_pkg_.rfind("-");
+            std::string name = s_pkg_.substr(0, pos_2);
+            std::string version = s_pkg_.substr(pos_2 + 1, s_pkg_.size());
+
+            specs::PackageInfo pkg_info{ name, version, build_string, channel };
+            return pkg_info;
+        }
+
+        PackageDiff PackageDiff::get_revision_pkg_diff(
+            std::vector<History::UserRequest> user_requests,
+            int target_revision
+        )
+        {
+            struct revision
+            {
+                int key = -1;
+                std::map<std::string, specs::PackageInfo> removed_pkg = {};
+                std::map<std::string, specs::PackageInfo> installed_pkg = {};
+            };
+
+            std::vector<revision> revisions;
+            for (auto r : user_requests)
+            {
+                if ((r.link_dists.size() > 0) || (r.unlink_dists.size() > 0))
+                {
+                    if (r.revision_num > target_revision)
+                    {
+                        revision rev{ /*.key = */ r.revision_num };
+                        for (auto ud : r.unlink_dists)
+                        {
+                            auto pkg_info = pkg_info_builder(ud);
+                            auto name = pkg_info.name;
+                            rev.removed_pkg[name] = std::move(pkg_info);
+                        }
+                        for (auto ld : r.link_dists)
+                        {
+                            auto pkg_info = pkg_info_builder(ld);
+                            auto name = pkg_info.name;
+                            rev.installed_pkg[name] = std::move(pkg_info);
+                        }
+                        revisions.push_back(rev);
+                    }
+                }
+            }
+
+            PackageDiff pkg_diff{};
+
+            auto handle_install = [&pkg_diff](revision& rev, const std::string& pkg_name)
```

Comment:
minor improvment: local lambdas that act as sub-functions can be `const` as their call function is already `const` (except if marked `mutable` but it's not the case here.